packages/authorization-server/src/services/tokenService.ts

```diff
@@ -0,0 +1,434 @@
+            consumerId: clientKidPurposeEntry.consumerId,
+            publicKey: clientKidPurposeEntry.publicKey,
+            algorithm: "RS256" /* TODO pass this as a parameter? */,
+            clientKind: clientKindTokenStates.consumer, // TODO this doesn't work with clientKidPurpose Entry.clientKind, but it should be already validated in the "when"
+            purposeState: {
+              state: clientKidPurposeEntry.purposeState,
+              versionId: clientKidPurposeEntry.purposeVersionId,
+            },
+            agreementId: clientKidPurposeEntry.agreementId,
+            agreementState: {
+              state: clientKidPurposeEntry.agreementState,
+            },
+            // TODO: make function for this split
+            eServiceId: unsafeBrandId<EServiceId>(
+              clientKidPurposeEntry.GSIPK_eserviceId_descriptorId.split("#")[0]
+            ),
+            eServiceState: {
+              state: clientKidPurposeEntry.descriptorState,
+              descriptorId: unsafeBrandId<DescriptorId>(
+                clientKidPurposeEntry.GSIPK_eserviceId_descriptorId.split(
+                  "#"
+                )[1]
+              ),
+              audience: clientKidPurposeEntry.descriptorAudience,
+              voucherLifespan: clientKidPurposeEntry.descriptorVoucherLifespan,
+            },
+          };
+          return key;
+        }
+      )
+      .when(
+        (entry) =>
+          entry.clientKind === clientKindTokenStates.consumer &&
+          entry.PK.startsWith(clientKidPrefix),
+        (entry) => {
+          throw keyTypeMismatch(clientKidPrefix, entry.clientKind);
+        }
+      )
+      .when(
+        (entry) =>
+          entry.clientKind === clientKindTokenStates.api &&
+          entry.PK.startsWith(clientKidPurposePrefix),
+        (entry) => {
+          throw keyTypeMismatch(clientKidPurposePrefix, entry.clientKind);
+        }
+      )
+      .when(
+        (entry) =>
+          entry.clientKind === clientKindTokenStates.api &&
+          entry.PK.startsWith(clientKidPrefix),
+        () => {
+          const clientKidEntry =
+            tokenGenerationEntry.data as TokenGenerationStatesClientEntry;
+
+          const key: ApiKey = {
+            kid: clientKidEntry.GSIPK_kid,
+            clientId: clientKidEntry.GSIPK_clientId,
+            consumerId: clientKidEntry.consumerId,
+            publicKey: clientKidEntry.publicKey,
+            algorithm: "RS256", // TODO pass this as a parameter?,
+            clientKind: clientKindTokenStates.api, // TODO this doesn't work with clientKidEntry.clientKind, but it should be already validated in the "when"
+          };
+          return key;
+        }
+      )
+      .otherwise(() => {
+        throw unexpectedTokenGenerationStatesEntry();
+      });
+  }
+  // } catch (error) {
+  //   // error handling.
+  //   // TODO Handle both dynamodb errors and throw error for empty public key
+  //   throw keyRetrievalFailed();
+  // }
+};
```

Comment:
Related to [this](https://github.com/pagopa/interop-be-monorepo/pull/1086/files#r1821201097)